File: wisp/transport/progress.py
"""Progress tracking for agent turns — pure data, no I/O.

Tracks phase transitions, tool execution counts, file changes,
and turn-level statistics for CLI rendering.
"""

from __future__ import annotations

from typing import Any

import json
import time
from dataclasses import dataclass, field

from wisp.core.events import AgentEvent, EventType

# ── Phase constants ─────────────────────────────────────────────

UNDERSTAND = "understand"
PLAN = "plan"
EXECUTE = "execute"
VERIFY = "verify"

_PHASE_ORDER = {UNDERSTAND: 0, PLAN: 1, EXECUTE: 2, VERIFY: 3}

_WRITE_TOOLS = frozenset({"write_file", "edit_file", "edit_file_multi"})
_EXECUTE_TOOLS = _WRITE_TOOLS | {"run_bash"}
_VERIFY_TOOLS = frozenset({"run_tests", "lsp_diagnostics", "diagnose"})
_PLAN_TOOLS = frozenset({"plan_task"})


@dataclass
class TurnProgress:
    """Snapshot of progress within a single turn."""

    turn_number: int = 0
    phase: str = UNDERSTAND
    tools_run: int = 0
    tools_succeeded: int = 0
    tools_failed: int = 0
    files_changed: list[str] = field(default_factory=list)
    start_time: float = 0.0
    current_tool: str | None = None
    current_tool_args: dict = field(default_factory=dict)
    current_tool_start: float = 0.0


class ProgressTracker:
    """Tracks agent progress across a turn from event stream.

    Pure data — no I/O, no side effects. Callers read ``progress``
    and ``on_event`` return values to drive rendering.
    """

    def __init__(self) -> None:
        self.progress = TurnProgress()
        self._has_written: bool = False
        self._seen_files: set[str] = set()

    # ── Lifecycle ───────────────────────────────────────────────

    def start_turn(self, turn_number: int) -> None:
        """Reset state for a new turn."""
        self.progress = TurnProgress(
            turn_number=turn_number,
            start_time=time.monotonic(),
        )
        self._has_written = False
        self._seen_files.clear()

    @property
    def elapsed(self) -> float:
        """Seconds since turn started."""
        if self.progress.start_time == 0:
            return 0.0
        return time.monotonic() - self.progress.start_time

    def on_done(self) -> dict:
        """Return final stats dict for the turn."""
        p = self.progress
        return {
            "turn_number": p.turn_number,
            "phase": p.phase,
            "tools_run": p.tools_run,
            "tools_succeeded": p.tools_succeeded,
            "tools_failed": p.tools_failed,
            "files_changed": list(p.files_changed),
            "elapsed": self.elapsed,
        }

    # ── Event dispatch ──────────────────────────────────────────

    def on_event(self, event: AgentEvent) -> str | None:
        """Process an event and return new phase if it changed."""
        etype = event.type if isinstance(event.type, str) else str(event.type)

        if etype == "tool_call":
            name = event.data.get("name", "")
            args = event.data.get("arguments", {})
            self.on_tool_call(name, args)
            return self._maybe_advance_phase(name)

        if etype == "tool_result":
            name = event.data.get("name", "")
            result = event.data.get("result", "")
            duration_ms = event.data.get("duration_ms")
            self.on_tool_result(name, result, duration_ms)
            return None

        if etype == "thinking":
            text = event.data.get("text", "")
            if len(text) > 500 and self.progress.phase == UNDERSTAND:
                old = self.progress.phase
                self.progress.phase = PLAN
                if old != PLAN:
                    return PLAN
            return None

        return None

    def on_tool_call(self, name: str, args: dict | None = None) -> None:
        """Record a tool invocation with its arguments."""
        self.progress.tools_run += 1
        self.progress.current_tool = name
        self.progress.current_tool_args = args or {}
        self.progress.current_tool_start = time.monotonic()

    def on_tool_result(
        self, name: str, result: Any, duration_ms: float | None = None
    ) -> None:
        """Record tool completion and track file changes from saved args."""
        self.progress.current_tool = None
        self._classify_result(result)
        self._track_files(name, self.progress.current_tool_args)
        self.progress.current_tool_args = {}

    # ── Internal ────────────────────────────────────────────────

    def _maybe_advance_phase(self, tool_name: str) -> str | None:
        """Check if tool call triggers a phase transition."""
        current = _PHASE_ORDER.get(self.progress.phase, 0)

        if tool_name in _PLAN_TOOLS and current < _PHASE_ORDER[PLAN]:
            self.progress.phase = PLAN
            return PLAN

        if tool_name in _EXECUTE_TOOLS:
            self._has_written = True
            if current < _PHASE_ORDER[EXECUTE]:
                self.progress.phase = EXECUTE
                return EXECUTE

        if tool_name in _VERIFY_TOOLS:
            target = VERIFY if self._has_written else current
            target_idx = _PHASE_ORDER.get(target, 0)
            if current < target_idx:
                self.progress.phase = target
                return target

        return None

    def _classify_result(self, result: Any) -> None:
        """Classify tool result as success or failure."""
        is_error = False
        if isinstance(result, dict):
            is_error = result.get("status") == "error"
        elif isinstance(result, str):
            try:
                parsed = json.loads(result)
                if isinstance(parsed, dict):
                    is_error = parsed.get("status") == "error"
            except (json.JSONDecodeError, TypeError):
                is_error = result.startswith("Error") or result.startswith("[")
        # Non-string, non-dict results (bytes, int, etc.) = success

        if is_error:
            self.progress.tools_failed += 1
        else:
            self.progress.tools_succeeded += 1

    def _track_files(self, name: str, args: dict) -> None:
        """Track file paths from write/edit tools."""
        if name not in _WRITE_TOOLS:
            return
        path = args.get("path", "")
        if path and path not in self._seen_files:
            self._seen_files.add(path)
            self.progress.files_changed.append(path)

File: wisp-ts/src/transport/progress.ts
/** Progress tracking for agent turns — pure data, no I/O.
 * Tracks phase transitions, tool execution counts, file changes,
 * and turn-level statistics for CLI rendering.
 */

import { AgentEvent } from "../core/events.js";

export const UNDERSTAND = "understand";
export const PLAN = "plan";
export const EXECUTE = "execute";
export const VERIFY = "verify";

const PHASE_ORDER: Record<string, number> = {
  [UNDERSTAND]: 0,
  [PLAN]: 1,
  [EXECUTE]: 2,
  [VERIFY]: 3,
};

const WRITE_TOOLS = new Set(["write_file", "edit_file", "edit_file_multi"]);
const EXECUTE_TOOLS = new Set([...WRITE_TOOLS, "run_bash"]);
const VERIFY_TOOLS = new Set(["run_tests", "lsp_diagnostics", "diagnose"]);
const PLAN_TOOLS = new Set(["plan_task"]);

export interface TurnProgressSnapshot {
  turnNumber: number;
  phase: string;
  toolsRun: number;
  toolsSucceeded: number;
  toolsFailed: number;
  filesChanged: string[];
  elapsed: number;
}

export class TurnProgress {
  turnNumber = 0;
  phase = UNDERSTAND;
  toolsRun = 0;
  toolsSucceeded = 0;
  toolsFailed = 0;
  filesChanged: string[] = [];
  startTime = 0;
  currentTool: string | null = null;
  currentToolArgs: Record<string, unknown> = {};
  currentToolStart = 0;
}

export class ProgressTracker {
  progress = new TurnProgress();
  private _hasWritten = false;
  private _seenFiles = new Set<string>();

  startTurn(turnNumber: number): void {
    this.progress = new TurnProgress();
    this.progress.turnNumber = turnNumber;
    this.progress.startTime = performance.now();
    this._hasWritten = false;
    this._seenFiles.clear();
  }

  get elapsed(): number {
    if (this.progress.startTime === 0) return 0;
    return (performance.now() - this.progress.startTime) / 1000;
  }

  onDone(): TurnProgressSnapshot {
    const p = this.progress;
    return {
      turnNumber: p.turnNumber,
      phase: p.phase,
      toolsRun: p.toolsRun,
      toolsSucceeded: p.toolsSucceeded,
      toolsFailed: p.toolsFailed,
      filesChanged: [...p.filesChanged],
      elapsed: this.elapsed,
    };
  }

  onEvent(event: AgentEvent): string | null {
    const etype = typeof event.type === "string" ? event.type : String(event.type);

    if (etype === "tool_call") {
      const name = typeof event.data.name === "string" ? event.data.name : "";
      const args = (event.data.arguments as Record<string, unknown>) ?? {};
      this.onToolCall(name, args);
      return this._maybeAdvancePhase(name);
    }

    if (etype === "tool_result") {
      const name = typeof event.data.name === "string" ? event.data.name : "";
      const result = event.data.result;
      const durationMs = typeof event.data.duration_ms === "number" ? event.data.duration_ms : undefined;
      this.onToolResult(name, result, durationMs);
      return null;
    }

    if (etype === "thinking") {
      const text = typeof event.data.text === "string" ? event.data.text : "";
      if (text.length > 500 && this.progress.phase === UNDERSTAND) {
        this.progress.phase = PLAN;
        return PLAN;
      }
      return null;
    }

    return null;
  }

  onToolCall(name: string, args?: Record<string, unknown>): void {
    this.progress.toolsRun += 1;
    this.progress.currentTool = name;
    this.progress.currentToolArgs = args ?? {};
    this.progress.currentToolStart = performance.now();
  }

  onToolResult(name: string, result: unknown, _durationMs?: number): void {
    this.progress.currentTool = null;
    this._classifyResult(result);
    this._trackFiles(name, this.progress.currentToolArgs);
    this.progress.currentToolArgs = {};
  }

  private _maybeAdvancePhase(toolName: string): string | null {
    const current = PHASE_ORDER[this.progress.phase] ?? 0;

    if (PLAN_TOOLS.has(toolName) && current < PHASE_ORDER[PLAN]) {
      this.progress.phase = PLAN;
      return PLAN;
    }

    if (EXECUTE_TOOLS.has(toolName)) {
      this._hasWritten = true;
      if (current < PHASE_ORDER[EXECUTE]) {
        this.progress.phase = EXECUTE;
      }
    }

    if (VERIFY_TOOLS.has(toolName) && current < PHASE_ORDER[VERIFY]) {
      this.progress.phase = VERIFY;
      return VERIFY;
    }

    return null;
  }

  private _classifyResult(result: unknown): void {
    if (
      result &&
      typeof result === "object" &&
      (result as Record<string, unknown>).status === "error"
    ) {
      this.progress.toolsFailed += 1;
    } else {
      this.progress.toolsSucceeded += 1;
    }
  }

  private _trackFiles(toolName: string, args: Record<string, unknown>): void {
    if (!WRITE_TOOLS.has(toolName)) return;
    const pathArg = args.path ?? args.filepath ?? args.file;
    if (typeof pathArg === "string") {
      if (!this._seenFiles.has(pathArg)) {
        this._seenFiles.add(pathArg);
        this.progress.filesChanged.push(pathArg);
      }
    }
  }
}
